niworkflows/utils/spaces.py

```diff
@@ -0,0 +1,418 @@
+    """Whether this space is standard or not."""
+    dim: int = attr.ib(default=3, repr=False)
+    """Dimensionality of the sampling manifold."""
+
+    def __attrs_post_init__(self):
+        """Extract cohort out of spec."""
+        if self.spec is None:
+            object.__setattr__(self, "spec", {})
+
+        if self.name.startswith('fsaverage'):
+            name = self.name
+            object.__setattr__(self, "name", "fsaverage")
+
+            if 'den' not in self.spec or name != "fsaverage":
+                spec = self.spec.copy()
+                spec['den'] = FSAVERAGE_DENSITY[name]
+                object.__setattr__(self, "spec", spec)
+
+        if self.name.startswith('fs'):
+            object.__setattr__(self, "dim", 2)
+
+        if self.name in self._standard_spaces:
+            object.__setattr__(self, "standard", True)
+
+        if self.spec and 'cohort' in self.spec:
+            spec = self.spec.copy()
+            value = spec.pop('cohort')
+            self._check_cohort('cohort', value)
+            object.__setattr__(self, "cohort", value)
+            object.__setattr__(self, "spec", spec)
+            return
+
+        _cohorts = _tfapi.TF_LAYOUT.get_cohorts(template=self.name)
+        if _cohorts:
+            _cohorts = ', '.join(['"cohort-%s"' % c for c in _cohorts])
+            raise ValueError(
+                'standard space "%s" is not fully defined.\n'
+                'Set a valid cohort selector from: %s.' % (self.name, _cohorts))
+
+    @name.validator
+    def _check_name(self, attribute, value):
+        if value.startswith('fsaverage'):
+            return
+        valid = list(self._standard_spaces) + NONSTANDARD_REFERENCES
+        if value not in valid:
+            raise ValueError(
+                'space identifier "%s" is invalid.\nValid '
+                'identifiers are: %s' % (value, ', '.join(valid)))
+
+    @cohort.validator
+    def _check_cohort(self, attribute, value):
+        valid = ['%s' % c for c in _tfapi.TF_LAYOUT.get_cohorts(
+            template=self.name)]
+        if value is not None and str(value) not in valid:
+            raise ValueError(
+                'standard space "%s" does not contain any cohort '
+                'named "%s".' % (self.name, value))
+
+    @spec.validator
+    def _check_spec(self, attribute, value):
+        if value is not None and not isinstance(value, dict):
+            raise TypeError(
+                "invalid space specification: %s." % str(value))
+
+    @classmethod
+    def from_string(cls, value):
+        """Create a new Space from string."""
+        _args = value.split(':')
+        spec = defaultdict(list, {})
+        for modifier in _args[1:]:
+            mitems = modifier.split('-', 1)
+            spec[mitems[0]].append(len(mitems) == 1 or mitems[1])
+
+        allspecs = _expand_entities(spec)
+
+        return [cls(_args[0], s) for s in allspecs]
+
+
+class SpatialReferences:
```

Comment:
playing off `Space.from_string()`, a class method could be implemented to process the `opt.output_spaces` input

Reply:
Working on this on the sMRIPrep side.